src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -387,12 +406,27 @@ private async ValueTask<ProtocolToken> ReceiveBlobAsync<TIOAdapter>(TIOAdapter a
             }
             else if (_lastFrame.Header.Type == TlsContentType.Handshake)
             {
-                if (_handshakeBuffer.ActiveReadOnlySpan[TlsFrameHelper.HeaderSize] == (byte)TlsHandshakeType.ClientHello &&
+                TlsFrameHelper.ProcessingOptions options = NetEventSource.IsEnabled ?
+                    TlsFrameHelper.ProcessingOptions.ServerName : TlsFrameHelper.ProcessingOptions.All;
+
+                if (NetEventSource.IsEnabled ||
+                    _handshakeBuffer.ActiveReadOnlySpan[TlsFrameHelper.HeaderSize] == (byte)TlsHandshakeType.ClientHello &&
                     _sslAuthenticationOptions!.ServerCertSelectionDelegate != null)
                 {
                     // Process SNI from Client Hello message
-                    TlsFrameHelper.TryGetHandshakeInfo(_handshakeBuffer.ActiveReadOnlySpan, ref _lastFrame);
-                    _sslAuthenticationOptions.TargetHost = _lastFrame.TargetName;
+                    TlsFrameHelper.TryGetHandshakeInfo(_handshakeBuffer.ActiveReadOnlySpan, ref _lastFrame, options);
```

Comment:
Let say we successfully parse SNI and have string with name. But we fail for what ever other reason. Should we not use it? The old SNI function would simply return string or null and in the null case SNI callback would still be called with empty string. That allows to to pick sime default certificate or fail if either side wishes.  

Maybe it would be cleaner if we split the tracing from the production code. This was attempt to do both in single pass. Maybe I can add more comment and perhaps tracing if TryGetHandshakeInfo returns false.

Reply:
I added result check and NetEventSource.Error for failure.